### DISCUSSION TITLE
"update everything every time" is a terrible policy

### DISCUSSION THREAD

**Discussion by FlorinAndrei** [Homebrew/discussions#578] [Everyday usage]
Today I had to do an emergency install with the socat package, to troubleshoot a system that was failing. I was under extreme time pressure. I ran "brew install socat".

To my shock and dismay, brew decided to upgrade the whole freaking collection of apps, while installing socat. That was... terrible.

Please don't do this. By default, the app should not take misguided initiative like this. If and when I need to upgrade the apps, I will do so myself.

Thanks.

**Comment by SMillerDev** [comment]
If your system is up to date this will never happen. See https://docs.brew.sh/FAQ#why-does-brew-upgrade-formula-also-upgrade-a-bunch-of-other-stuff

**Comment by carlocab** [comment]
> "update everything every time" is a terrible policy

It might be, but that's not Homebrew's upgrade/install policy. It will only upgrade software that needs upgrading unless specifically requested with an unqualified `brew upgrade`.

**Comment by tlk** [comment]
Agree. Consider disabling homebrew auto updates:
```bash
echo "export HOMEBREW_NO_AUTO_UPDATE=1" >> ~/.zshrc
```

I believe this restores old-school behaviour of the package manager and puts you in charge of running `brew update` to fetch Formula updates.

**Reply by FlorinAndrei** [reply]
> Agree. Consider disabling homebrew auto updates:
> 
> ```shell
> echo "export HOMEBREW_NO_AUTO_UPDATE=1" >> ~/.zshrc
> ```
> 
> I believe this restores old-school behaviour of the package manager and puts you in charge of running `brew update` to fetch Formula updates.

I will do that, thanks. I cannot afford to wait for the whole freakin' starship to boot up and launch every time I need to install a two-bit utility.

Heck, even "yum install" is better in this regard.

I miss the elegant simplicity of old Brew.

**Reply by iMichka** [reply]
> old Brew

You mean; before 2016 brew? The autoupdate feature on install was added there: https://github.com/Homebrew/brew/pull/679

Back then we had a big amount of issues being opened because there was no autoupdate:
- users got confused as why they did not get the latest version of their software
- users who found bugs in software would open issues, and our stock answer was 'did you try brew update'?

This solution was a good way to get rid of the whole bug triaging which drained the energy of our maintainers, and overall a lot of users were really happy about the change.

**Comment by JakkuSakura** [comment]
I still enjoy the time with pacman and other pkg managers

**Comment by jamesbebbington** [comment]
Could homebrew not by default at least ask the user to confirm that they want to update a bunch of seemingly* unrelated packages updated before it does so? I believe I understand the technical reasons for why it sometimes needs to update the whole world and all of its canines, but from a UX perspective, the decision to default to this behaviour is more than a little perplexing.

*from a user's perspective

**Reply by carlocab** [reply]
Requiring zero user input for nearly all commands (the only exceptions I know of are some casks, but those are required by upstream installers) is a deliberate design decision. This is needed to allow `brew` to be used in scripts and CI, where there typically is no user to interact with. Requiring user input to do `brew upgrade` will break many scripts and CI workflows.

If you want to check what happens when you do `brew upgrade` before actually doing it, use `brew upgrade --dry-run`.

**Comment by webdeb** [comment]
There must be a better solution, like npm. Its not the best, but with pnpm and yarn it is getting pretty solid.
